src/Views/LocationPanel.vala

```diff
@@ -118,91 +122,92 @@ public class SecurityPrivacy.LocationPanel : Granite.SimpleSettingsPage {
         }
     }
 
-    private void populate_app_listbox () {
-        load_remembered_apps ();
-
-        foreach (var row in listbox.get_children ()) {
-            listbox.remove (row);
-        }
-
-        foreach (var app in remembered_apps) {
-            string app_id = app.get_child_value (0).get_string ();
-            bool authed = app.get_child_value (1).get_variant ().get_child_value (0).get_boolean ();
-            var app_info = new DesktopAppInfo (app_id + ".desktop");
-
-            var app_row = new LocationRow (app_info, authed);
+    private Gtk.Widget create_widget_func (Object object) {
+        var app_permission = (AppPermission) object;
+        var app_row = new LocationRow (app_permission);
+        app_row.notify["authed"].connect (() => {
+            string[] permissions = {app_row.authed ? "EXACT" : "NONE", app_row.timestamp};
+            try {
+                permission_store.set_permission (PERMISSIONS_TABLE, true, PERMISSIONS_ID, app_permission.id, permissions);
+            } catch (Error e) {
+                critical (e.message);
+            }
 
-            app_row.active_changed.connect ((active) => {
-                uint32 level = get_app_level (app_id);
-                save_app_settings (app_id, active, level);
-            });
-
-            listbox.add (app_row);
-        }
-    }
+        });
 
-    private void load_remembered_apps () {
-        remembered_apps = location_settings.get_value ("remembered-apps");
-        remembered_apps_dict = new VariantDict (location_settings.get_value ("remembered-apps"));
+        return app_row;
     }
 
-    private void save_app_settings (string desktop_id, bool authorized, uint32 accuracy_level) {
-        Variant[] tuple_vals = new Variant[2];
-        tuple_vals[0] = new Variant.boolean (authorized);
-        tuple_vals[1] = new Variant.uint32 (accuracy_level);
-        remembered_apps_dict.insert_value (desktop_id, new Variant.tuple (tuple_vals));
-        location_settings.set_value ("remembered-apps", remembered_apps_dict.end ());
-        load_remembered_apps ();
+    private void load_permissions () {
+        liststore.remove_all ();
+
+        try {
+            Variant permissions, data;
+            permission_store.lookup (PERMISSIONS_TABLE, PERMISSIONS_ID, out permissions, out data);
+
+            string app_id = "";
+            string[] app_permissions = {""};
+            var iter = permissions.iterator ();
+            while (iter.next ("{s^as}", ref app_id, ref app_permissions)) {
+                var app_permission = new AppPermission (
+                    app_id,
+                    app_permissions[0],
+                    app_permissions[1]
+                );
+
+                liststore.append (app_permission);
+            }
```

Comment:
Can you try with this
```suggestion
            unowned string app_id;
            unowned string[] app_permissions;
            var iter = permissions.iterator ();
            while (iter.next ("{s^as}", out app_id, out app_permissions)) {
                var app_permission = new AppPermission (
                    app_id,
                    app_permissions[0],
                    app_permissions[1]
                );

                liststore.append (app_permission);
            }
```

Reply:
(with the modified "{&s^a&s}" to allow less string allocations)

Reply:
Works for me! Thanks :)